### ISSUE TITLE
Cannot execute replicated DDL query, maximum retries exceeded

### ISSUE THREAD

**Issue by avointsev** [ClickHouse/ClickHouse#70472]
Hello , 

I become to get a strange behaviour with ALTER on cluster command
I have 2 nodes cluster with 1 shard  and a lot of small , but similar databases
I'm getting the next error on few tables (some of tables altered without errors)

```
ALTER TABLE sc_alfa2.users_authentication_logs ON CLUSTER sc
    (ADD COLUMN IF NOT EXISTS `fingerprint_v2` Nullable(String))

Query id: 50c1cb44-aca3-4944-9f7d-2bc7e994e7ad

   ┌─host──────────┬─port─┬─status─┬─error─────────────────────────────────────────────────────────┬─num_hosts_remaining─┬─num_hosts_active─┐
1. │ 172.18.16.156 │ 9000 │    341 │ Cannot execute replicated DDL query, maximum retries exceeded │                   1 │                0 │
2. │ 172.18.16.51  │ 9000 │    341 │ Cannot execute replicated DDL query, maximum retries exceeded │                   0 │                0 │
   └───────────────┴──────┴────────┴───────────────────────────────────────────────────────────────┴─────────────────────┴──────────────────┘
← Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.)
2 rows in set. Elapsed: 3.319 sec.

Received exception from server (version 24.8.4):
Code: 341. DB::Exception: Received from localhost:9000. DB::Exception: There was an error on [172.18.16.156:9000]: Cannot execute replicated DDL query, maximum retries exceeded. (UNFINISHED)
```
Table statement
```
statement: CREATE TABLE sc_alfa2.users_authentication_logs
(
    `login_at` DateTime,
    `user_id` UInt64,
    `username` String,
    `user_type` LowCardinality(String),
    `domain` String,
    `is_vr` UInt8,
    `login_method` LowCardinality(String),
    `user_agent` String,
    `device_type` LowCardinality(String),
    `fingerprint` String,
    `ip` String,
    `isp` String,
    `country` LowCardinality(String),
    `city` String,
    `is_activated_2fa` UInt8,
    `is_suspicious_login` UInt8,
    `suspicious_confirmation_type` LowCardinality(String),
    `suspicious_rule` String,
    `is_proxy_ip` UInt8,
    `proxy_ip_type` LowCardinality(String),
    `is_trusted_browser` UInt8,
    `trusted_browser_id` String,
    `is_2fa_requested` UInt8
)
ENGINE = ReplicatedMergeTree('/var/lib/zookeeper/sc_alfa2/{shard}/users_authentication_logs', '{replica}')
PARTITION BY toYYYYMM(login_at)
ORDER BY (login_at, user_id)
TTL login_at + toIntervalYear(2)
SETTINGS index_granularity = 8192

1 row in set. Elapsed: 0.003 sec.
```

Early I got this problem at  23.4.2  version. Then I updated to last LTS 24.8.4.13 , but the problem still here. 

I'm getting next events  the clickhouse.err.log 

```
2024.10.08 08:23:08.276171 [ 317947 ] {50c1cb44-aca3-4944-9f7d-2bc7e994e7ad} <Error> executeQuery: Code: 341. DB::Exception: There was an error on [172.18.16.156:9000]: Cannot execute replicated DDL query, maximum retries exceeded. (UNFINISHED) (version 24.8.4.13 (official build)) (from 127.0.0.1:48752) (in query: ALTER TABLE sc_alfa2.users_authentication_logs ON CLUSTER
 sc ADD COLUMN IF NOT EXISTS `fingerprint_v2` Nullable(String) ;), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000daffc3b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007e59fcc
2. DB::DDLQueryStatusSource::generate() @ 0x0000000011e13b97
3. DB::ISource::tryGenerate() @ 0x000000001303219b
4. DB::ISource::work() @ 0x0000000013031ea7
5. DB::ExecutionThreadContext::executeTask() @ 0x000000001304b0e7
6. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000001303f9b0
7. DB::PipelineExecutor::execute(unsigned long, bool) @ 0x000000001303ee42
8. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true, true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineE
xecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x
000000001304ebaa
9. void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std
::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000dbd7509
10. start_thread @ 0x00000000000081ca
11. clone @ 0x0000000000039e73

2024.10.08 08:23:08.276360 [ 317947 ] {50c1cb44-aca3-4944-9f7d-2bc7e994e7ad} <Error> TCPHandler: Code: 341. DB::Exception: There was an error on [172.18.16.156:9000]: Cannot execute replicat
ed DDL query, maximum retries exceeded. (UNFINISHED), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000daffc3b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007e59fcc
2. DB::DDLQueryStatusSource::generate() @ 0x0000000011e13b97
3. DB::ISource::tryGenerate() @ 0x000000001303219b
4. DB::ISource::work() @ 0x0000000013031ea7
5. DB::ExecutionThreadContext::executeTask() @ 0x000000001304b0e7
6. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000001303f9b0
7. DB::PipelineExecutor::execute(unsigned long, bool) @ 0x000000001303ee42
8. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true, true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineE
xecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x
000000001304ebaa
9. void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std
::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000dbd7509
10. start_thread @ 0x00000000000081ca
11. clone @ 0x0000000000039e73
```

I've checked  ZK paths ( ZK parameters looks fine at first view ) and applied `system restart` . 
What I can do more for debug this problem ?

I've attached clickhouse server config (yml)
[server.yml.txt](https://github.com/user-attachments/files/17291058/server.yml.txt)

**Comment by avointsev** [comment]
[server.yml.txt](https://github.com/user-attachments/files/17291018/server.yml.txt)

**Comment by den-crane** [comment]
duplicate https://github.com/ClickHouse/ClickHouse/issues/54902#issuecomment-2088301845

**Comment by avointsev** [comment]
Without on cluster I'm getting next error 
```
Received exception from server (version 24.8.4):
Code: 517. DB::Exception: Received from localhost:9000. DB::Exception: Metadata on replica is not up to date with common metadata in Zookeeper. It means that this replica still not applied some of previous alters. Probably too many alters executing concurrently (highly not recommended). You can retry this error. (CANNOT_ASSIGN_ALTER)
```

**Comment by avointsev** [comment]
I was given a solution in Clickhouse telegram channel https://github.com/ClickHouse/ClickHouse/issues/54902#issuecomment-2088301845

The solution is works. But it looks like a workaround , not a solution of the problem .
And the main reason of the problem that metadata in Zookeeper and Clickhouse are different
Why it happened ? It means that it can repeat tomorrow or in any day again .

**Comment by avointsev** [comment]
workaround

-- clickhouse  any node
```
sc-clickhouse-hotel-02 :) SELECT * FROM system.zookeeper
WHERE path like '/var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/%'
AND name = 'metadata'
SETTINGS allow_unrestricted_reads_from_keeper = 'true';

SELECT *
FROM system.zookeeper
WHERE (path LIKE '/var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/%') AND (name = 'metadata')
SETTINGS allow_unrestricted_reads_from_keeper = 'true'

Query id: 525fd55d-6ca3-4e86-8808-b771b9df6c45

   ┌─name─────┬─value──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬─path──────────────────────────────────────────────────────────────────────────────────┐
1. │ metadata │ metadata format version: 1
date column:
sampling expression:
index granularity: 8192
mode: 0
sign column:
primary key: login_at, user_id
data format version: 1
partition key: toYYYYMM(login_at)
sorting key: login_at, user_id
ttl: login_at + toIntervalYear(2)
granularity bytes: 10485760
 │ /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.51 │
   └──────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────────────────────────────────────────────────────────────┘
   ┌─name─────┬─value──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬─path───────────────────────────────────────────────────────────────────────────────────┐
2. │ metadata │ metadata format version: 1
date column:
sampling expression:
index granularity: 8192
mode: 0
sign column:
primary key: login_at, user_id
data format version: 1
partition key: toYYYYMM(login_at)
sorting key: login_at, user_id
ttl: login_at + toIntervalYear(2)
granularity bytes: 10485760
 │ /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.156 │
   └──────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────┘

2 rows in set. Elapsed: 26.897 sec. Processed 356.06 thousand rows, 81.95 MB (13.24 thousand rows/s., 3.05 MB/s.)
Peak memory usage: 186.00 MiB.


[zk: localhost:2181(CONNECTED) 17] get /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.156/metadata_version
0
[zk: localhost:2181(CONNECTED) 18] get /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.51/metadata_version
0
```

-- zookeper leader node
```
[zk: localhost:2181(CONNECTED) 17] get /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.156/metadata_version
0
[zk: localhost:2181(CONNECTED) 18] get /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.51/metadata_version
0

[zk: localhost:2181(CONNECTED) 10] set /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.51/metadata_version '1'
[zk: localhost:2181(CONNECTED) 11] set /var/lib/zookeeper/sc_alfa1/01/users_authentication_logs/replicas/172.18.16.156/metadata_version '1'
```

-- clickhouse  any node
```
sc-clickhouse-hotel-02 :) system restart replica on cluster sc sc_alfa1.users_authentication_logs ;

SYSTEM RESTART REPLICA ON CLUSTER sc sc_alfa1.users_authentication_logs

Query id: 6e4beb9d-4008-4a79-8d3e-f5d5df19e0fc

   ┌─host──────────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
1. │ 172.18.16.156 │ 9000 │      0 │       │                   1 │                0 │
2. │ 172.18.16.51  │ 9000 │      0 │       │                   0 │                0 │
   └───────────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

2 rows in set. Elapsed: 0.068 sec.


sc-clickhouse-hotel-02 :) ALTER TABLE sc_alfa1.users_authentication_logs ON CLUSTER sc ADD COLUMN IF NOT EXISTS `fingerprint_v2` Nullable(String) ;

ALTER TABLE sc_alfa1.users_authentication_logs ON CLUSTER sc
    (ADD COLUMN IF NOT EXISTS `fingerprint_v2` Nullable(String))

Query id: b922942c-4087-40a9-8938-e33cb0e1b3f2

   ┌─host──────────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
1. │ 172.18.16.156 │ 9000 │      0 │       │                   1 │                0 │
2. │ 172.18.16.51  │ 9000 │      0 │       │                   0 │                0 │
   └───────────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

2 rows in set. Elapsed: 0.061 sec.


sc-clickhouse-hotel-02 :) desc sc_alfa1.users_authentication_logs ;

DESCRIBE TABLE sc_alfa1.users_authentication_logs

Query id: 3f11f7e1-90ac-424b-800a-715d73f528cf

    ┌─name─────────────────────────┬─type───────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
 1. │ login_at                     │ DateTime               │              │                    │         │                  │                │
 2. │ user_id                      │ UInt64                 │              │                    │         │                  │                │
 3. │ username                     │ String                 │              │                    │         │                  │                │
 4. │ user_type                    │ LowCardinality(String) │              │                    │         │                  │                │
 5. │ domain                       │ String                 │              │                    │         │                  │                │
 6. │ is_vr                        │ UInt8                  │              │                    │         │                  │                │
 7. │ login_method                 │ LowCardinality(String) │              │                    │         │                  │                │
 8. │ user_agent                   │ String                 │              │                    │         │                  │                │
 9. │ device_type                  │ LowCardinality(String) │              │                    │         │                  │                │
10. │ fingerprint                  │ String                 │              │                    │         │                  │                │
11. │ ip                           │ String                 │              │                    │         │                  │                │
12. │ isp                          │ String                 │              │                    │         │                  │                │
13. │ country                      │ LowCardinality(String) │              │                    │         │                  │                │
14. │ city                         │ String                 │              │                    │         │                  │                │
15. │ is_activated_2fa             │ UInt8                  │              │                    │         │                  │                │
16. │ is_suspicious_login          │ UInt8                  │              │                    │         │                  │                │
17. │ suspicious_confirmation_type │ LowCardinality(String) │              │                    │         │                  │                │
18. │ suspicious_rule              │ String                 │              │                    │         │                  │                │
19. │ is_proxy_ip                  │ UInt8                  │              │                    │         │                  │                │
20. │ proxy_ip_type                │ LowCardinality(String) │              │                    │         │                  │                │
21. │ is_trusted_browser           │ UInt8                  │              │                    │         │                  │                │
22. │ trusted_browser_id           │ String                 │              │                    │         │                  │                │
23. │ is_2fa_requested             │ UInt8                  │              │                    │         │                  │                │
24. │ fingerprint_v2               │ Nullable(String)       │              │                    │         │                  │                │
    └──────────────────────────────┴────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

24 rows in set. Elapsed: 0.001 sec.
```

**Comment by den-crane** [comment]
>Why it happened ? It means that it can repeat tomorrow or in any day again .

Because of the bug in Clickhouse server it could happen when you added a replica and Clickhouse version is >=23.8 and <= 23.12.
The bug was fixed in 24.1, but consequences of the bug have to be fixed manually.

----

There is one more fix for consequences:

v24.9.1.3278-stable.md:* After [this issue](https://github.com/ClickHouse/ClickHouse/pull/59946#issuecomment-1943653197) there are quite a few table replicas in production such that their `metadata_version` node value is both equal to `0` and is different from the respective table's `metadata` node version. This leads to `alter` queries failing on such replicas. [#69274](https://github.com/ClickHouse/ClickHouse/pull/69274) ([Miсhael Stetsyuk](https://github.com/mstetsyuk)).

But it fixes only if node value is both equal to `0` -> you never altered a table before.

**Comment by avointsev** [comment]
thanks
